### PR TITLE
Fix spy(..., marker=<not-None>, origin="lower")

### LINKED ISSUE #16252: axes.spy plotting function doesn't respect origin='lower' kwarg when marker is not None
### Bug report

**Bug summary**

`spy` always draws with the origin in the upper left corner if `marker` is not `None`, even if `origin='lower'` is provided as a keyword argument.

**Code for reproduction**

```python
points = np.zeros([10, 5])
points[5:] = 1
fig, ax = plt.subplots()
ax.spy(points, marker='.', origin='lower')

```

**Actual outcome**
spy_point_marker

After this point, the origin appears to be fixed in the upper left corner. Any subsequent commands (`imshow`, `spy`) seem to ignore `origin='lower'`.

**Expected outcome**

Note that `ax.spy(points, origin='lower')` (omitting the `marker` argument) works properly:
spy_default_marker

Though it's worth pointing out that the x-axis ticks and labels always end up in upper spine, not the lower one, regardless of `origin='upper'` or `origin='lower'`. The behavior of `imshow` is a bit different. If `origin='lower'`, the ticks and labels are moved to the bottom.

The documentation for spy notes that:
```
If *Z* is a `scipy.sparse.spmatrix` or *marker* or *markersize* are
*None*, a `~matplotlib.lines.Line2D` object will be returned with
the value of marker determining the marker type, and any
remaining kwargs passed to `~.Axes.plot`.
```
So that seems to explain why including `marker` changes the function's behavior.

**Matplotlib version**

  * Operating system: Centos 7
  * Matplotlib version: 3.1.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.6.8
  * Jupyter version (if applicable):
  * Other libraries: 

Installed using anaconda.

### PR DESCRIPTION
**Pull request by anntzer**
and validate the origin kwarg both for spy() and for imshow().

There are some other slightly questionable choices in spy() (e.g. always
calling `xaxis.set_ticks_position("both")` but not doing the same for
the yaxis), but let's leave this as is for now.

Closes #16252.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -7560,6 +7555,11 @@ def spy(self, Z, precision=0, marker=None, markersize=None,
               non-square pixels.
             - *None*: Use :rc:`image.aspect`.
 
+        origin : {'upper', 'lower'}, default: :rc:`image.origin`
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: this just makes the doc order match the signature order.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -7618,23 +7619,27 @@ def spy(self, Z, precision=0, marker=None, markersize=None,
...
-            self.set_ylim(nr - 0.5, -0.5)
+            if origin == "upper":
+                self.set_ylim(nr - 0.5, -0.5)
+            else:
+                self.set_ylim(-0.5, nr - 0.5)
             self.set_aspect(aspect)
-            ret = marks
         self.title.set_y(1.05)
-        self.xaxis.tick_top()
+        if origin == "upper":
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: This seems like an additional change in behavior?

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: that was explicitly mentioned in the report ("Though it's worth pointing out that the x-axis ticks and labels always end up in upper spine, not the lower one, regardless of origin='upper' or origin='lower'. The behavior of imshow is a bit different. If origin='lower', the ticks and labels are moved to the bottom.") and clearly looks like an error as well.

Comment (tacaswell) [reply lib/matplotlib/axes/_axes.py]: :+1:

**Pull request merged**